File: Week_03/kotlin/src/_98_ValidateBinarySearchTree.kt
class _98_Morris {
    fun isValidBST(root: TreeNode?): Boolean {
        var node = root
        var prev = Long.MIN_VALUE
        while (node != null) {
            if (node.left == null) {
                if (node.`val` <= prev) return false
                prev = node.`val`.toLong()
                node = node.right
            } else {
                var predecessor = node.left!!
                while (predecessor.right != null && predecessor.right != node) {
                    predecessor = predecessor.right!!
                }
                if (predecessor.right == null) {
                    predecessor.right = node
                    node = node.left
                } else {
                    predecessor.right = null
                    if (node.`val` <= prev) return false
                    prev = node.`val`.toLong()
                    node = node.right
                }
            }
        }
        return true
    }
}

File: Week_03/cpp/src/_98_ValidateBinarySearchTree.cpp
#include <iostream>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution
{
public:
    bool isValidBST(TreeNode *root)
    {
        long long prev = LONG_MIN;
        return dfs(root, &prev);
    }

    bool dfs(TreeNode *root, long long *prev)
    {
        if (!root)
            return true;
        if (!dfs(root->left, prev))
            return false;
        if (root->val <= *prev)
            return false;
        *prev = root->val;
        return dfs(root->right, prev);
    }
};